### ISSUE TITLE
NetworkSecurityConfig - Peer not trusted by any of the system trust managers.

### ISSUE THREAD

**Issue by adriancretu** [koush/AndroidAsync#653]
OS: Android 9.0
Issue: If no network security config provided, everything works and SSL is enforced by Android 9.0. However, if I want to allow clear-text traffic (the default before Android 9.0), but at the same time enforce SSL for a domain, things break, though they shouldn't. This seems to happen simply because there's a "domain-config" entry, which requires a different approach to verify.
WSS connection fails when using NetworkSecurityConfig XML like this:

`network_security_config.xml`

```
<network-security-config>
    <base-config cleartextTrafficPermitted="true">
        <trust-anchors>
            <certificates src="system" />
        </trust-anchors>
    </base-config>

    <domain-config cleartextTrafficPermitted="false">
        <domain includeSubdomains="false">{exact_domain_name}</domain>
    </domain-config>
</network-security-config>
```

Exception:
```
com.koushikdutta.async.AsyncSSLException: Peer not trusted by any of the system trust managers.
        at com.koushikdutta.async.AsyncSSLSocketWrapper.handleHandshakeStatus(AsyncSSLSocketWrapper.java:322)
        at com.koushikdutta.async.AsyncSSLSocketWrapper.write(AsyncSSLSocketWrapper.java:401)
        at com.koushikdutta.async.AsyncSSLSocketWrapper.handleHandshakeStatus(AsyncSSLSocketWrapper.java:276)
        at com.koushikdutta.async.AsyncSSLSocketWrapper.access$000(AsyncSSLSocketWrapper.java:30)
        at com.koushikdutta.async.AsyncSSLSocketWrapper$5.onDataAvailable(AsyncSSLSocketWrapper.java:213)
        at com.koushikdutta.async.Util.emitAllData(Util.java:23)
        at com.koushikdutta.async.AsyncNetworkSocket.onReadable(AsyncNetworkSocket.java:152)
        at com.koushikdutta.async.AsyncServer.runLoop(AsyncServer.java:821)
        at com.koushikdutta.async.AsyncServer.run(AsyncServer.java:658)
        at com.koushikdutta.async.AsyncServer.access$800(AsyncServer.java:44)
        at com.koushikdutta.async.AsyncServer$14.run(AsyncServer.java:600)
     Caused by: java.security.cert.CertificateException: Domain specific configurations require that hostname aware checkServerTrusted(X509Certificate[], String, String) is used
        at android.security.net.config.RootTrustManager.checkServerTrusted(RootTrustManager.java:111)
        at com.koushikdutta.async.AsyncSSLSocketWrapper.handleHandshakeStatus(AsyncSSLSocketWrapper.java:298)
```

If I remove the `<domain-config>` entry, the connection works. Actually, it doesn;t really matter what domains I actually add to the list, because `android.security.net.config.RootTrustManager.checkServerTrusted` actually just checks whether there's an existing domains section or not.

**Comment by toyajohnae301** [comment]
Stop sending me junk emails

**Comment by lukenowakcom** [comment]
Same issue, did anyone found solution?

**Comment by zkrige** [comment]
Same issue. I'm having to do a variation of this just to make it work
```
Ion.getDefault(ctx).getHttpClient().getSSLSocketMiddleware().setTrustManagers(new TrustManager[] {new X509TrustManager() {
            @Override
            public void checkClientTrusted(final X509Certificate[] chain, final String authType) throws CertificateException {}

            @Override
            public void checkServerTrusted(final X509Certificate[] chain, final String authType) throws CertificateException {}

            @Override
            public X509Certificate[] getAcceptedIssuers() {
                return new X509Certificate[0];
            }
        }});
```

**Comment by eduardobvale** [comment]
Same issue here, @zkrige solution is working, but i'm afraid of being too insecure. Trusting any certificates seems to be insecure. Anyone got updates on that?

**Comment by mengoni** [comment]
is there any solution available for AndroidAsync (not Ion)?

**Comment by koush** [comment]
I believe I have a fix for this and will release an update.
The SSLContext has a trust manager, and then another root trust manager is being consulted as well (which is in error). The second trust manager is the one that is throwing the error.

**Comment by koush** [comment]
This commit contains the fix if someone can test it.

https://github.com/koush/AndroidAsync/commit/5adb6e0e2caeb6991fec7faed265a39dadd0f7ab

**Comment by silentnuke** [comment]
@koush confirmed that it fixed issue, thank you! Can you publish new version?

**Comment by koush** [comment]
Fixed and released in 3.0.9.
